File: 0201-0300/0214-Shortest Palindrome/0214-Shortest Palindrome.cpp
class Solution {
public:
    string shortestPalindrome(string s) {
        string r(s);
        reverse(r.begin(), r.end());
        string t = s + "#" + r;
        vector<int> lps(t.size());
        int len = 0, i = 1;
        while (i < t.size()) {
            if (t[i] == t[len]) {
                lps[i++] = ++len;
            }
            else if (len != 0) {
                len = lps[len - 1];
            }
            else {
                ++i;
            }
        }
        
        return r + s.substr(len);
    }
};

File: 0201-0300/0214-Shortest Palindrome/0214-Shortest Palindrome.py
class Solution:
    def shortestPalindrome(self, s: str) -> str:
        temp = s + '#' + s[::-1]
        i = 1
        l = 0
        lps = [0] * len(temp)
        while i < len(temp):
            if temp[i] == temp[l]:
                lps[i] = l + 1
                i += 1
                l += 1
            elif l != 0:
                l = lps[l - 1]
            else:
                i += 1
        return s[l:][::-1] + s
